modules/gapi/src/backends/onnx/gonnxbackend.cpp

```diff
@@ -181,17 +181,21 @@ inline void preprocess(const cv::Mat& src,
     } else {
         // 8U input: full preprocessing path
         GAPI_Assert(src.depth()   == CV_8U && "Only 8U data type is supported for preproc");
-        GAPI_Assert(ti.dims.size() == 4u && "Only NCHW/NHWC layouts are supported for preproc");
+        GAPI_Assert((ti.dims.size() == 4u || ti.dims.size() == 3u)
+                    && "Only NCHW/NHWC/CHW/HWC layouts are supported for preproc");
+
+        const bool with_batch = ti.dims.size() == 4u ? true : false;
+        const int shift = with_batch ? 0 : 1;
 
         const auto ddepth = toCV(ti.type);
         GAPI_Assert((ddepth == CV_8U || ddepth == CV_32F)
                     && "Only 8U and 32F model input is supported for 8U data");
 
         // Assess the expected input layout
         const bool is_hwc = [&](int ch) {
-            if (ti.is_grayscale)       return false; // 1,1,h,w
-            else if (ti.dims[3] == ch) return true;  // _,_,_,c
-            else if (ti.dims[1] == ch) return false; // _,c,_,_
+            if (ti.is_grayscale)       return false; // ?,1,h,w
```

Comment:
Comment alignment is not yet fixed.

Reply:
`return false` moved right. If comment about it.